### PR TITLE
ARROW-17450: [C++][Parquet] Add support for uint8 boolean decode in addition to bool array

### PR DESCRIPTION
**Pull request by sfc-gh-nthimmegowda**
Commit [466018084861f86a9171803c6d689e9c1c1efb5a](https://github.com/apache/arrow/commit/466018084861f86a9171803c6d689e9c1c1efb5a) added support for RLE boolean decoder. We refactored some additional code making it streamlined with other cases for decoder. 

However there was a downstream dependency for a Decode function which taken in an array of `uint8` instead of `bool`. To not break any existing workload, adding back support for decode boolean datatype with array of `uint8`

### CODE REVIEW

File: cpp/src/parquet/encoding.h [hunk 1]
```diff
@@ -394,6 +394,12 @@ class DictDecoder : virtual public TypedDecoder<DType> {
 // ----------------------------------------------------------------------
 // TypedEncoder specializations, traits, and factory functions
 
+class BooleanDecoder : virtual public TypedDecoder<BooleanType> {
+ public:
+  using TypedDecoder<BooleanType>::Decode;
+  virtual int Decode(uint8_t* buffer, int max_values) = 0;
```
Comment (pitrou) [inline cpp/src/parquet/encoding.h:400]: Can you add a docstring? It should be clear that `buffer` will contain bit-packed values.

Comment (sfc-gh-nthimmegowda) [reply cpp/src/parquet/encoding.h:408]: Done. Added a docstring.

File: cpp/src/parquet/encoding.h [hunk 2]
```diff
@@ -394,6 +394,20 @@ class DictDecoder : virtual public TypedDecoder<DType> {
 // ----------------------------------------------------------------------
 // TypedEncoder specializations, traits, and factory functions
 
+class BooleanDecoder : virtual public TypedDecoder<BooleanType> {
+ public:
+  using TypedDecoder<BooleanType>::Decode;
+
+  /// \brief Decode values into a buffer
```
Comment (pitrou) [inline cpp/src/parquet/encoding.h:401]:
```suggestion
  /// \brief Decode and bit-pack values into a buffer
```

File: cpp/src/parquet/encoding.h [hunk 3]
```diff
@@ -394,6 +394,20 @@ class DictDecoder : virtual public TypedDecoder<DType> {
...
+  ///
+  /// \param[in] buffer destination for decoded values
+  /// This buffer contains bit-packed values.
```
Comment (pitrou) [inline cpp/src/parquet/encoding.h:404]:
```suggestion
  /// This buffer will contain bit-packed values.
```

**Review by pitrou — APPROVED** [review]

**Pull request merged**